File: src/catch_them_all_py/catch_them_all_py/turtle_controller.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from turtlesim.msg import Pose
from geometry_msgs.msg import Twist
from my_robot_interfaces.msg import AliveTurtles
from my_robot_interfaces.msg import Turtle
from my_robot_interfaces.srv import CatchTurtle
from my_robot_interfaces.srv import TunePid
from my_robot_interfaces.srv import GetPidValues
from functools import partial

import math
from simple_pid import PID

class turtleControllerNode(Node):
    def __init__(self):
        super().__init__("turtle_controller")
    
        self.declare_parameter("catch_closest_turtle_first", True)
        self.catch_closest_turtle_first = self.get_parameter("catch_closest_turtle_first").value
        self.pose_ = None
        self.pose_sub = self.create_subscription(Pose, "turtle1/pose", self.callbackTurtlePose,10)
        self.alive_turtles_sub = self.create_subscription(AliveTurtles, "alive_turtles", self.updateTarget,10)
        self.cmd_Vel_pub = self.create_publisher(Twist, "turtle1/cmd_vel", 10)
        self.tune_pid_server = self.create_service(TunePid, "tune_pid", self.callbackTunePid)
        self.get_pid_server = self.create_service(GetPidValues, "get_pid_values", self.callbackGetPidValues)
        self.control_timer = self.create_timer(0.001,self.callbackControllLoop)

        self.turtue_to_catch = None

        self.pid_lin_vel = PID(2.0, 0.0, 0.0, 0.0)
        self.pid_ang_vel = PID(10.0, 0.0, 0.0, 0.0)
        
        

        self.get_logger().info("turtle_controller has been started")
        
    def updateTarget(self, msg:AliveTurtles):
        if len(msg.trutles) >0:
            closest_turtle = None
            closest_distance = None
            if self.catch_closest_turtle_first:
                for turtle in msg.trutles:
                    dist_x = turtle.x - self.pose_.x
                    dist_y = turtle.y - self.pose_.y
                    distance = math.sqrt(dist_x*dist_x + dist_y*dist_y)
                    if closest_turtle == None or distance < closest_distance:
                        closest_turtle = turtle
                        closest_distance = distance
                        
                self.turtue_to_catch = closest_turtle
            else:
                self.turtue_to_catch = msg.trutles[0]

    def callbackTurtlePose(self,msg):
        self.pose_ = msg
        
    def callbackControllLoop(self):
        if self.pose_ == None or self.turtue_to_catch == None:
            return
        msg = Twist()
        
        dist_x = self.turtue_to_catch.x - self.pose_.x
        dist_y = self.turtue_to_catch.y  - self.pose_.y
        dist = math.sqrt(dist_x * dist_x + dist_y * dist_y)
        
        if dist > 0.5:
            
            x_vel = self.pid_lin_vel(-dist)
            msg.linear.x = x_vel

        
            target_theta = math.atan2(dist_y, dist_x)
            diff_theta = target_theta - self.pose_.theta
                    
            clipped_theta = pi_clip(-diff_theta)
            
            msg.angular.z = self.pid_ang_vel(clipped_theta)
            
        else: #target reached
            msg.linear.x = 0.0
            msg.angular.z = 0.0
            self.callCatchTurtle (self.turtue_to_catch.name)
            self.turtue_to_catch = None
                    
        self.cmd_Vel_pub.publish(msg)
        
    def callCatchTurtle(self, turtle_name):
        client = self.create_client(CatchTurtle, "catch_turtle")

        while not client.wait_for_service(1.0):
            self.get_logger().warn("Waiting for Server....")

        request = CatchTurtle.Request()
        
        request.name = turtle_name
        
        
        future = client.call_async(request)
        future.add_done_callback(partial(self.callback_call_turtle_catch, turtle_name=request.name))
    
    def callback_call_turtle_catch(self, future, turtle_name):
        try:
            response = future.result()
            if not response.success:
                self.get_logger().info(f"Turtle {turtle_name} could not be caught")
                
        except Exception as e:
            self.get_logger().error(f"Servce call faild with exception {e}") 

    def callbackTunePid(self, request:TunePid.Request, response:TunePid.Response):
        self.pid_lin_vel.Kp = request.linear_velocity.p
        self.pid_lin_vel.Ki = request.linear_velocity.i
        self.pid_lin_vel.Kd = request.linear_velocity.d
        
        self.pid_ang_vel.Kp = request.angular_velocity.p
        self.pid_ang_vel.Ki = request.angular_velocity.i
        self.pid_ang_vel.Kd = request.angular_velocity.d
       
        response.success = True
        return response
    
    def callbackGetPidValues(self, request:GetPidValues.Request, response:GetPidValues.Response):
        response.linear_velocity.p = self.pid_lin_vel.Kp
        response.linear_velocity.i = self.pid_lin_vel.Ki
        response.linear_velocity.d = self.pid_lin_vel.Kd
        
        response.angular_velocity.p = self.pid_ang_vel.Kp
        response.angular_velocity.i = self.pid_ang_vel.Ki
        response.angular_velocity.d = self.pid_ang_vel.Kd
        
        return response


def pi_clip(angle):
    if angle > 0:
        if angle > math.pi:
            return angle - 2*math.pi
    else:
        if angle < -math.pi:
            return angle + 2*math.pi
    return angle


        
    

def main(args=None):
    rclpy.init(args=args)
    node = turtleControllerNode()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == "__main__":
    main()
    

File: src/catch_them_all_cpp/src/turtle_controller.cpp
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "turtlesim/msg/pose.hpp"
#include "my_robot_interfaces/msg/alive_turtles.hpp"
#include "my_robot_interfaces/srv/catch_turtle.hpp"

class turtleControllNode : public rclcpp::Node
{
public:
    turtleControllNode() : Node("turtle_controller"), name_("turtle1"), turtlesim_up_(false)
    {
        this->declare_parameter("catch_closest_turtle_first", true);
        catch_closest_turtle_first_ = this->get_parameter("catch_closest_turtle_first").as_bool();

        cmd_Vel_pub = this->create_publisher<geometry_msgs::msg::Twist>(
            name_ + "/cmd_vel", 10);
        pose_sub = this->create_subscription<turtlesim::msg::Pose>(
            name_ + "/pose", 10, std::bind(&turtleControllNode::callbackPosse, this, std::placeholders::_1));
        alive_turtles_sub = this->create_subscription<my_robot_interfaces::msg::AliveTurtles>(
            "alive_turtles", 10, std::bind(&turtleControllNode::callbackTurtleToCatch, this, std::placeholders::_1));

        control_timer = this->create_wall_timer(std::chrono::milliseconds(10), std::bind(&turtleControllNode::controlLoop, this));

        RCLCPP_INFO(this->get_logger(), "turtle_controller has been started");
    }

private:
    void callbackPosse(const turtlesim::msg::Pose::SharedPtr pose)
    {
        pose_ = *pose.get();
        turtlesim_up_ = true;
    }

    double getDistanceFromCurrentPose(my_robot_interfaces::msg::Turtle turtle)
    {
        double dist_x = turtle.x - pose_.x;
        double dist_y = turtle.y - pose_.y;

        return std::sqrt(dist_x * dist_x + dist_y * dist_y);
    }

    void callbackTurtleToCatch(const my_robot_interfaces::msg::AliveTurtles::SharedPtr msg)
    {
        if (!msg->trutles.empty())
        {
            if (catch_closest_turtle_first_)
            {
                my_robot_interfaces::msg::Turtle closest_turtle = msg->trutles.at(0);
                double closest__turtle_diistance = getDistanceFromCurrentPose(closest_turtle);

                for (int i = 1; i < (int)msg->trutles.size(); i++)
                {
                    double distance = getDistanceFromCurrentPose(msg->trutles.at(i));
                    if (distance < closest__turtle_diistance)
                    {
                        closest_turtle = msg->trutles.at(i);
                        closest__turtle_diistance = distance;
                    }
                }

                turtle_to_catch_ = closest_turtle;
            }
            else
            {
                turtle_to_catch_ = msg->trutles.at(0);
            }
        }
    }

    void publlisshVmdVel(double x, double theta)
    {
        auto msg = geometry_msgs::msg::Twist();
        msg.linear.x = x;
        msg.angular.z = theta;
        cmd_Vel_pub->publish(msg);
    }

    void controlLoop()
    {
        if (!turtlesim_up_ || turtle_to_catch_.name == "")
        {
            return;
        }

        double dist_x = turtle_to_catch_.x - pose_.x;
        double dist_y = turtle_to_catch_.y - pose_.y;
        double distance = std::sqrt(dist_x * dist_x + dist_y * dist_y);

        auto msg = geometry_msgs::msg::Twist();

        if (distance > 0.5)
        {
            msg.linear.x = 2 * distance;

            double steering_angle = std::atan2(dist_y, dist_x);
            double angle_diff = steering_angle - pose_.theta;
            if (angle_diff > M_PI)
            {
                angle_diff -= 2 * M_PI;
            }
            else if (angle_diff < -M_PI)
            {
                angle_diff += 2 * M_PI;
            }
            msg.angular.z = 6 * angle_diff;
        }
        else
        {
            msg.linear.x = 0.0;
            msg.angular.z = 0.0;

            catch_turtle_thread_.push_back(std::make_shared<std::thread>(std::bind(&turtleControllNode::callCatchTurtleService, this, turtle_to_catch_.name)));
            turtle_to_catch_.name = "";
        }
        cmd_Vel_pub->publish(msg);
    }

    void callCatchTurtleService(std::string turtle_name)
    {
        auto client = this->create_client<my_robot_interfaces::srv::CatchTurtle>("catch_turtle");
        while (!client->wait_for_service(std::chrono::seconds(1)))
        {
            RCLCPP_WARN(this->get_logger(), "Waiting for catch_turtle Server to be up...");
        }

        auto request = std::make_shared<my_robot_interfaces::srv::CatchTurtle_Request>();
        request->name = turtle_name;

        auto future = client->async_send_request(request);

        try
        {
            auto resopnse = future.get();
            if (!resopnse->success)
            {
                RCLCPP_ERROR(this->get_logger(), "failed to catch turtle");
                
            }
            
        }
        catch(const std::exception& e)
        {
            RCLCPP_ERROR(this->get_logger(), "Service call failed. ");
            
        }
        
        
    }


    std::string name_;
    turtlesim::msg::Pose pose_;
    bool turtlesim_up_;
    bool catch_closest_turtle_first_;
    my_robot_interfaces::msg::Turtle turtle_to_catch_;

    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr cmd_Vel_pub;
    rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr pose_sub;
    rclcpp::Subscription<my_robot_interfaces::msg::AliveTurtles>::SharedPtr alive_turtles_sub;
    rclcpp::TimerBase::SharedPtr control_timer;

    std::vector<std::shared_ptr<std::thread>> catch_turtle_thread_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<turtleControllNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}